> @RaduNichita, in this PR, I focused on replacing `stdint.h` types with Unikraft defined, but I dug up the origin of the variables and I came up with the conclusion that the `unsigned long` and `unsigned` can be replaced with `__u64` and `__u8`. I will change this as well.

In principle, I would agree with @rares-miculescu here, the scope of this work is to replace libc types. On the specific changes:

`get_promiscuous` is the implementation of the netdev op defined in `netdev_core.h` which returns `unsigned`: https://github.com/unikraft/unikraft/blob/staging/lib/uknetdev/include/uk/netdev_core.h#L349. Therefore, the original definition was correct. In fact, the updated version generates a warning:
```
/home/mpp/devel/unikraft/uk_helloworld/unikraft/drivers/virtio/net/virtio_net.c:1258:28: warning: initialization of ‘unsigned int (*)(struct uk_netdev *)’ from incompatible pointer type ‘__u8 (*)(struct uk_netdev *)’ {aka ‘unsigned char (*)(struct uk_netdev *)’} [-Wincompatible-pointer-types]
 1258 |         .promiscuous_get = virtio_net_promisc_get,
      |                            ^~~~~~~~~~~~~~~~~~~~~~
/home/mpp/devel/unikraft/uk_helloworld/unikraft/drivers/virtio/net/virtio_net.c:1258:28: note: (near initialization for ‘virtio_netdev_ops.promiscuous_get’)
```
Please revert this change.

Regarding the cast in `virtio_pci_handle()`: `pci_isr_addr` is already of type `__u64` so the cast in redundant. I don't understand the purpose of the double cast (i.e. `virtio_cread8((void *)(unsigned long)d->pci_isr_addr, 0)`) before this change was applied, as casting `_u64` to `void *` would already suffice. So you can either remove the second cast, or leave it as it was.

--- a/drivers/virtio/9p/virtio_9p.c
+++ b/drivers/virtio/9p/virtio_9p.c
@@ -60,7 +60,7 @@ struct virtio_9p_device {
 	/* Virtqueue reference. */
 	struct virtqueue *vq;
 	/* Hw queue identifier. */
-	uint16_t hwvq_id;
+	__u16 hwvq_id;
 	/* libuk9p associated device (NULL if the device is not in use). */
 	struct uk_9pdev *p9dev;
 	/* Scatter-gather list. */
@@ -133,7 +133,7 @@ static int virtio_9p_request(struct uk_9pdev *p9dev,
 	struct virtio_9p_device *dev;
 	int rc, host_notified = 0;
 	unsigned long flags;
-	size_t read_segs, write_segs;
+	__sz read_segs, write_segs;
 	bool failed = false;
 
 	UK_ASSERT(p9dev);
@@ -173,7 +173,7 @@ static int virtio_9p_request(struct uk_9pdev *p9dev,
 	}
 
 	if (req->recv.zc_buf) {
-		uint32_t recv_size = req->recv.size + req->recv.zc_size;
+		__u32 recv_size = req->recv.size + req->recv.zc_size;
 
 		rc = uk_sglist_append(&dev->sg, req->recv.zc_buf,
 				req->recv.zc_size);
@@ -184,7 +184,7 @@ static int virtio_9p_request(struct uk_9pdev *p9dev,
 
 		/* Make eure there is sufficient space for Rerror replies. */
 		if (recv_size < UK_9P_RERROR_MAXSIZE) {
-			uint32_t leftover = UK_9P_RERROR_MAXSIZE - recv_size;
+			__u32 leftover = UK_9P_RERROR_MAXSIZE - recv_size;
 
 			rc = uk_sglist_append(&dev->sg,
 					req->recv.buf + recv_size, leftover);
@@ -235,7 +235,7 @@ static int virtio_9p_recv(struct virtqueue *vq, void *priv)
 {
 	struct virtio_9p_device *dev;
 	struct uk_9preq *req = NULL;
-	uint32_t len;
+	__u32 len;
 	int rc = 0;
 	int handled = 0;
 

--- a/drivers/virtio/blk/virtio_blk.c
+++ b/drivers/virtio/blk/virtio_blk.c
@@ -95,15 +95,15 @@ struct uk_blkdev_queue {
 	struct virtqueue *vq;
 	/* The libukblkdev queue identifier */
 	/* It is also the virtqueue identifier */
-	uint16_t lqueue_id;
+	__u16 lqueue_id;
 	/* Allocator */
 	struct uk_alloc *a;
 	/* The nr. of descriptor limit */
-	uint16_t max_nb_desc;
+	__u16 max_nb_desc;
 	/* The nr. of descriptor user configured */
-	uint16_t nb_desc;
+	__u16 nb_desc;
 	/* The flag to interrupt on the queue */
-	uint8_t intr_enabled;
+	__u8 intr_enabled;
 	/* Reference to virtio_blk_device  */
 	struct virtio_blk_device *vbd;
 	/* The scatter list and its associated fragments */
@@ -120,7 +120,7 @@ struct virtio_blkdev_request {
 	struct uk_blkreq *req;
 	struct uk_list_head free_list_head;
 	struct virtio_blk_outhdr virtio_blk_outhdr;
-	uint8_t status;
+	__u8 status;
 };
 
 static int virtio_blkdev_request_set_sglist(struct uk_blkdev_queue *queue,
@@ -130,11 +130,11 @@ static int virtio_blkdev_request_set_sglist(struct uk_blkdev_queue *queue,
 {
 	struct virtio_blk_device *vbdev;
 	struct uk_blkreq *req;
-	size_t data_size = 0;
-	size_t segment_size;
-	size_t segment_max_size;
-	size_t idx;
-	uintptr_t start_data;
+	__sz data_size = 0;
+	__sz segment_size;
+	__sz segment_max_size;
+	__sz idx;
+	__uptr start_data;
 	int rc = 0;
 
 	UK_ASSERT(queue);
@@ -142,7 +142,7 @@ static int virtio_blkdev_request_set_sglist(struct uk_blkdev_queue *queue,
 
 	req = virtio_blk_req->req;
 	vbdev = queue->vbd;
-	start_data = (uintptr_t)req->aio_buf;
+	start_data = (__uptr)req->aio_buf;
 	data_size = req->nb_sectors * sector_size;
 	segment_max_size = vbdev->max_size_segment;
 
@@ -174,7 +174,7 @@ static int virtio_blkdev_request_set_sglist(struct uk_blkdev_queue *queue,
 		}
 
 	rc = uk_sglist_append(&queue->sg, &virtio_blk_req->status,
-			sizeof(uint8_t));
+			sizeof(__u8));
 	if (unlikely(rc != 0)) {
 		uk_pr_err("Failed to append to sg list %d\n", rc);
 		goto out;
@@ -496,9 +496,9 @@ static int virtio_blkdev_queue_intr_disable(struct uk_blkdev *dev __unused,
  * This function setup the vring infrastructure.
  */
 static int virtio_blkdev_vqueue_setup(struct uk_blkdev_queue *queue,
-		uint16_t nr_desc)
+		__u16 nr_desc)
 {
-	uint16_t max_desc;
+	__u16 max_desc;
 	struct virtqueue *vq;
 
 	UK_ASSERT(queue);
@@ -533,8 +533,8 @@ static int virtio_blkdev_vqueue_setup(struct uk_blkdev_queue *queue,
 }
 
 static struct uk_blkdev_queue *virtio_blkdev_queue_setup(struct uk_blkdev *dev,
-		uint16_t queue_id,
-		uint16_t nb_desc,
+		__u16 queue_id,
+		__u16 nb_desc,
 		const struct uk_blkdev_queue_conf *queue_conf)
 {
 	struct virtio_blk_device *vbdev;
@@ -603,7 +603,7 @@ static int virtio_blkdev_queue_release(struct uk_blkdev *dev,
 }
 
 static int virtio_blkdev_queue_info_get(struct uk_blkdev *dev,
-		uint16_t queue_id,
+		__u16 queue_id,
 		struct uk_blkdev_queue_info *qinfo)
 {
 	struct virtio_blk_device *vbdev = NULL;
@@ -633,7 +633,7 @@ static int virtio_blkdev_queues_alloc(struct virtio_blk_device *vbdev,
 				    const struct uk_blkdev_conf *conf)
 {
 	int rc = 0;
-	uint16_t i = 0;
+	__u16 i = 0;
 	int vq_avail = 0;
 	__u16 qdesc_size[conf->nb_queues];
 
@@ -714,7 +714,7 @@ static int virtio_blkdev_start(struct uk_blkdev *dev)
 static int virtio_blkdev_stop(struct uk_blkdev *dev)
 {
 	struct virtio_blk_device *d;
-	uint16_t q_id;
+	__u16 q_id;
 	int rc = 0;
 
 	UK_ASSERT(dev != NULL);

--- a/drivers/virtio/mmio/virtio_mmio.c
+++ b/drivers/virtio/mmio/virtio_mmio.c
@@ -61,8 +61,8 @@
 
 static struct uk_alloc *a;
 struct virtio_mmio_device_id {
-	uint16_t device_id;
-	uint32_t vendor;
+	__u16 device_id;
+	__u32 vendor;
 };
 
 struct virtio_mmio_device {

--- a/drivers/virtio/mmio/virtio_mmio_cmdl.c
+++ b/drivers/virtio/mmio/virtio_mmio_cmdl.c
@@ -24,7 +24,7 @@ UK_LIBPARAM_PARAM_ARR_ALIAS(device, uk_libparam_devices, charp,
 			    MAX_DEV_CMDLINE, "virtio-mmio devices");
 
 /* Helper to parse numeric string with size suffix (eg "4K") into numeric */
-static char *parse_size(const char *size_str, size_t *sz)
+static char *parse_size(const char *size_str, __sz *sz)
 {
 	char *endptr;
 

